Framework/PythonInterface/mantid/api/src/Exports/IFunction.cpp

```diff
@@ -35,6 +36,32 @@ PyObject *getCategories(IFunction &self) {
   return registered;
 }
 
+/// An object for constructing a shared_ptr that won't ever delete its pointee
+class NoDeleting {
+public:
+  /// Does nothing
+  void operator()(void *) {}
+  /// Does nothing
+  void operator()(const void *) {}
+};
+
+//------------------------------------------------------------------------------------------------------
+/**
+* Something that makes IFunction return to python a composite function
+* for Product function, Convolution or
+* any similar superclass of composite function.
+* @param self :: Enables it to be called as a member function on the
+* IFunction class
+*/
+Mantid::API::CompositeFunction_sptr castToCompositeFunction(IFunction &self) {
```

Comment:
For composite functions, I would access the component functions _only_ by their composite index. For instance:
```
c1 = f1 + f2
c2 = f3 + f4
cc = c1 + c2
```
then `cc[1][0] == f3`. I think the adoption of prefixes like "f" or numerical suffixes like in "Gaussian 0" are somewhat arbitrary. The composite index [1][0] is a more natural representation of the hierarchical nature of the composite function. This even allows for fancy slicing. For non-composite functions, `f[0]` would return itself. In this convention, `__len__` would make sense as the number of functions, not the number of parameters.
Also, I would not use numerical indexes to access fitting parameter. I think they should be accessed only by their names.